File: xray-frontend/src/api/nodes.ts
import { api } from "./client";

/** Интерфейс объекта Ноды */
export interface Node {
  id: number;
  name: string;
  address: string;
  api_url: string;
  reality_server_address: string;
  is_active: boolean;
  
  // Версионность конфига
  desired_version: number;
  applied_version: number;
  
  // Статус
  last_heartbeat: string | null; // Приходит строкой ISO
}

/** Тип для создания новой ноды */
export interface NodeCreate {
  name: string;
  address: string;
  api_url: string;
  reality_server_address: string;
  is_active?: boolean;
  secret_token?: string; // Опционально, так как сервер может сгенерировать сам
}

/** Тип для обновления ноды (все поля опциональны) */
export interface NodeUpdate extends Partial<NodeCreate> {}

/** Ответ при генерации нового токена */
export interface RotateTokenResponse {
  status: string;
  new_token: string;
  warning: string;
}

/** Получить список всех нод */
export const getNodes = async (): Promise<Node[]> => {
  const res = await api.get<Node[]>("/nodes/all");
  return res.data;
};

/** Зарегистрировать новую ноду */
export const registerNode = async (data: NodeCreate): Promise<Node> => {
  const res = await api.post<Node>("/nodes/register", data);
  return res.data;
};

/** Обновить данные ноды */
export const updateNode = async (id: number, data: NodeUpdate): Promise<Node> => {
  const res = await api.patch<Node>(`/nodes/update/${id}`, data);
  return res.data;
};

/** Удалить ноду */
export const deleteNode = async (id: number): Promise<{ status: string; message: string }> => {
  const res = await api.delete(`/nodes/delete/${id}`);
  return res.data;
};

/** Сгенерировать новый секретный токен для ноды */
export const rotateNodeToken = async (id: number): Promise<RotateTokenResponse> => {
  const res = await api.post<RotateTokenResponse>(`/nodes/${id}/rotate-token`);
  return res.data;
};

File: app/api/endpoints/nodes.py
# app/api/endpoints/nodes.py

from fastapi import APIRouter, Depends, HTTPException, Header
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List

from app.core.database import get_db
from app.core.dependencies import get_current_admin
from app.schemas.node import NodeCreate, NodeResponse, NodeConfigResponse, NodeUpdate 
from app.services.node_service import NodeService

router = APIRouter()

# --- АДМИНСКИЕ ЭНДПОИНТЫ ---

@router.post("/register", response_model=NodeResponse)
async def register_node(
    obj_in: NodeCreate, 
    db: AsyncSession = Depends(get_db),
    admin: dict = Depends(get_current_admin)
):
    service = NodeService(db)
    return await service.register_node(obj_in)

@router.get("/all", response_model=List[NodeResponse])
async def list_nodes(
    db: AsyncSession = Depends(get_db),
    admin: dict = Depends(get_current_admin)
):
    service = NodeService(db)
    return await service.get_all_nodes()

# --- ЭНДПОИНТЫ ДЛЯ АГЕНТОВ (НОД) ---

@router.get("/config/{node_id}", response_model=NodeConfigResponse)
async def get_node_config(
    node_id: int,
    x_node_token: str = Header(...), # Нода передает свой secret_token в заголовках
    db: AsyncSession = Depends(get_db)
):
    """Эндпоинт, куда нода 'стучится' за обновлением конфига"""
    service = NodeService(db)
    config = await service.get_latest_config(node_id, x_node_token)
    
    if not config:
        raise HTTPException(status_code=404, detail="Config not found for this node")
    
    # Обновляем время последнего визита ноды
    await service.update_heartbeat(node_id, applied_version=config.version)
    
    return config

@router.patch("/update/{node_id}", response_model=NodeResponse)
async def update_node(
    node_id: int,
    obj_in: NodeUpdate,
    db: AsyncSession = Depends(get_db),
    admin: dict = Depends(get_current_admin)
):
    service = NodeService(db)
    updated_node = await service.update_node(node_id, obj_in)
    if not updated_node:
        raise HTTPException(status_code=404, detail="Node not found")
    return updated_node

@router.delete("/delete/{node_id}")
async def delete_node(
    node_id: int,
    db: AsyncSession = Depends(get_db),
    admin: dict = Depends(get_current_admin)
):
    service = NodeService(db)
    success = await service.delete_node(node_id)
    if not success:
        raise HTTPException(status_code=404, detail="Node not found")
    return {"status": "success", "message": f"Node {node_id} deleted"}

@router.post("/{node_id}/rotate-token")
async def rotate_node_token(
    node_id: int,
    db: AsyncSession = Depends(get_db),
    admin: dict = Depends(get_current_admin)
):
    """Смена токена безопасности ноды"""
    service = NodeService(db)
    new_token = await service.rotate_node_token(node_id)
    
    if not new_token:
        raise HTTPException(status_code=404, detail="Node not found")
        
    return {
        "status": "success",
        "new_token": new_token,
        "warning": "Update the token in your Node Agent config immediately. Connection will be lost until updated."
    }